### DISCUSSION TITLE
Custom-UI with React - Add navitems & more

### DISCUSSION THREAD

**Discussion by egillben** [vendurehq/vendure#1474] [Q&A] [answered]
Hello

I have been reading the documentations and trying to wrap my head around using react as my choice for Admin-UI extensions.

So my question is this:  Is React-UI extension only available on routes? or is it possible to use for example "addActionBarItem" and "registerFormInputComponent" with react components?

And also an extra question: Is it possible to create unique index on custom fields on tables?

**Comment by michaelbromley** [comment, marked as answer]
Hi!

> Is React-UI extension only available on routes? or is it possible to use for example "addActionBarItem" and "registerFormInputComponent" with react components?

Correct - A React app can currently only be mounted at a given route. I think it _might_ be technically feasible to make some kind of wrapper that allows you to embed a React app as a ActionBar item or a FormInputComponent, but it would be a _lot_ of complexity, lots of added bundle size (a whole app just for a single component), and would be an inferior end result to using a native Angular component for those use cases. 

> And also an extra question: Is it possible to create unique index on custom fields on tables?

Not currently, no. Can you give a little more insight into the use case here? It might be something work considering supporting.

**Reply by egillben** [reply]
Thank you for your answer. We really like your product :) 

> Not currently, no. Can you give a little more insight into the use case here? It might be something work considering supporting.

We are developing a web store for a customer and we chose Vendure to do the job. The customer is using external booking system to store all customers and products and even shipping methods.  So we have created a sync plugin to fetch all the data and we are using custom field to store the ID from the external system. If the custom field could be unique we would use UPSERT when running the sync task but now we're just fetching all the items first and then checking if the external item exists in the database which results in a very expensive operations that could be simplified by using a unique column.

If the item exists in the database we update it, otherwise we create it.

**Reply by michaelbromley** [reply]
I'm glad you're enjoying working with Vendure!

And thanks for the extra info. It makes sense, and I just took a look at how we're handling custom field generation and I think it should be possible to support this. If you create a feature request issue I'll add it to the roadmap.
